Fix bug in random matrix generator and add a Benchmark

FlorentinD opened the pull request:
Resulting in the wrong row selection strategy being used, and hence being very very slow.

Corresponding issue: #98 

Other changes are:
- when selecting rows, they are directly written
- use boolean[] instead of BitSet
- use for-each instead for Streaming BitSet

Thanks @ennerf for the last two tips!

FlorentinD commented:
When trying to run the benchmark, I had to comment out some others like `WatchDoubleStepQR` as there are some `Nullable` compile errors.

Besides I was a bit surprised, that these aren't JMH benchmarks as they could provide better measures.

lessthanoptimal commented:
You stumbled into an older part of the code that hasn't been updated to JMH yet. `ejml-experimental` is where ideas were being tested out and is a bit of a repository of "this failed don't try it again" or does X slow things down or not. It was written before JMH existed or was popular.

The `Nullable` error is weird. Do you get it when you run `./gradlew build` ?  According to the NullAway rule in gradle that code should have been ignored since it was in a `benchmarks` directory.

lessthanoptimal commented on main/ejml-experimental/benchmarks/src/org/ejml/sparse/BenchmarkRandomMatrixGenerator.java:
```diff
@@ -0,0 +1,55 @@
+
+import java.util.Random;
+
+public class BenchmarkRandomMatrixGenerator {
```
Could you move this to `main/ejml-dsparse/benchmarks/src/org/ejml/sparse/csc` and JMH-ify it?

FlorentinD replied:
Sure, I just couldn't find this directory before :)

lessthanoptimal commented on main/ejml-experimental/benchmarks/src/org/ejml/sparse/BenchmarkRandomMatrixGenerator.java:
```diff
@@ -0,0 +1,55 @@
+
+import java.util.Random;
+
+public class BenchmarkRandomMatrixGenerator {
```
The name should be something like `BenchmarkRandomMatrices_DSCC`

FlorentinD commented:
As I built something inside the experimental module, I think it just threw than.
Normally running `./gradlew build` runs fine.

FlorentinD commented:
However I am also getting similar errors in `ejml-dsparse:compileBenchmarskJava`, when executing the main method of the benchmark.

I helped myself by disabling errorprune in the `build.gradle` for getting first jmh results.

First results:
```
Benchmark                                     (avgEntriesPerColumn)  (dimension)  Mode  Cnt     Score     Error  Units
BenchmarkRandomMatrices_DSCC.generateUniform                     10        10000  avgt   10     5.380 ±   0.176  ms/op
BenchmarkRandomMatrices_DSCC.generateUniform                     10        40000  avgt   10    50.859 ±   5.373  ms/op
BenchmarkRandomMatrices_DSCC.generateUniform                    300        10000  avgt   10   218.111 ±  17.141  ms/op
BenchmarkRandomMatrices_DSCC.generateUniform                    300        40000  avgt   10   885.351 ±  43.038  ms/op
BenchmarkRandomMatrices_DSCC.rectangle                           10        10000  avgt   10    87.043 ±   1.487  ms/op
BenchmarkRandomMatrices_DSCC.rectangle                           10        40000  avgt   10  1171.267 ±   8.975  ms/op
BenchmarkRandomMatrices_DSCC.rectangle                          300        10000  avgt   10   633.921 ±  75.517  ms/op
BenchmarkRandomMatrices_DSCC.rectangle                          300        40000  avgt   10  3595.355 ± 107.327  ms/op
```

Commit:
Fix bug in random matrix generator

Resulting in the wrong row selection strategy being used.

Also:
- Add a benchmark for random sparse matrix generation
- Suppress unused method warnings for parameterized tests methods

FlorentinD commented:
As a side note:
I shortly tried parallelizing the generation using `EjmlConcurrency.loopBlocks()`, but this was ca. 3x slower than the single threaded version. Potentially I need to have separate arrays per thread and merge them at the end. Although the threads write into exclusive parts of the arrays.

lessthanoptimal reviewed (approved):
LGTM

lessthanoptimal commented:
I'm curious why you're getting those errors and not me. Must have a different configuration in IntelliJ.

Threads:
I would need to see the code to provide specific comments. Did you have a different `Random` instance for each thread? I think (double check) it is thread safe by uses synchronize so you really don't want to use it across threads. If there's anything that grows an array that could be killing performance and should also be looked at carefully.

lessthanoptimal commented:
could you squash this commit too? Thanks.

FlorentinD commented:
No I didn't use different instances, this might be a cause right.

I will squash the commits tommorow evening :)

FlorentinD commented:
Squashed them :) 

Regarding the threads:
I will start another try at some point, but the current performance is sufficient for my use-case.

If its running fine on your machine and CI, than it must be my Intelij.
